idpartner/__init__.py

```diff
@@ -0,0 +1,234 @@
+import urllib.parse
+import json
+import uuid
+import os
+import secrets
+import hashlib
+import base64
+import requests
+from authlib.oauth2.rfc7636 import create_s256_code_challenge
+import authlib.jose as jose
+from datetime import datetime, timedelta
+
+
+class IDPartner:
+    SUPPORTED_AUTH_METHODS = [
+        "client_secret_basic",
+        "tls_client_auth",
+        "private_key_jwt",  # For backward compatibility
+    ]
+    SIGNING_ALG = "PS256"
+    ENCRYPTION_ALG = "RSA-OAEP"
+    ENCRYPTION_ENC = "A256CBC-HS512"
+
+    def __init__(self, oauth, config):
+        if not oauth:
+            raise ValueError("OAuth missing.")
+        if not config:
+            raise ValueError("Config missing.")
+
+        default_config = {
+            "account_selector_service_url": "https://auth-api.idpartner.com/oidc-proxy",
+            "token_endpoint_auth_method": "client_secret_basic",
+            "jwks": None,
+            "client_secret": None,
+        }
+
+        self.config = {**default_config, **config}
+
+        if (
+            self.config.get("token_endpoint_auth_method")
+            not in self.SUPPORTED_AUTH_METHODS
+        ):
+            raise ValueError(
+                f"Unsupported token_endpoint_auth_method '{self.config.get('token_endpoint_auth_method')}'. "
+                f"It must be one of ({', '.join(self.SUPPORTED_AUTH_METHODS)})"
+            )
+
+        client_secret_config = (
+            {"client_secret": self.config["client_secret"]}
+            if self.config.get("token_endpoint_auth_method") == "client_secret_basic"
+            else {}
+        )
+
+        jwks_config = (
+            {
+                "authorization_encrypted_response_alg": self.ENCRYPTION_ALG,
+                "authorization_encrypted_response_enc": self.ENCRYPTION_ENC,
+                "id_token_encrypted_response_alg": self.ENCRYPTION_ALG,
+                "id_token_encrypted_response_enc": self.ENCRYPTION_ENC,
+                "request_object_signing_alg": self.SIGNING_ALG,
+            }
+            if self.config.get("jwks")
+            else {}
+        )
+
+        self.config = {
+            **self.config,
+            **{
+                "authorization_signed_response_alg": self.SIGNING_ALG,
+                "id_token_signed_response_alg": self.SIGNING_ALG,
+            },
+            **client_secret_config,
+            **jwks_config,
+        }
+        if self.config.get("jwks"):
+            self.config["jwks"] = jose.JsonWebKey.import_key_set(self.config["jwks"])
+        self.oauth = oauth
+
+    def generate_proofs(self):
+        return {
+            "state": generate_token(43),
+            "nonce": generate_token(43),
+            "code_verifier": generate_token(43),
+        }
+
+    def jwks(self):
+        if not self.config.get("jwks"):
+            return {}
+        public_jwks = []
+        for jwk in self.config.get("jwks").keys:
+            public_jwk = jwk.as_dict()
```

Comment:
This would include private claims too, rigth? i think it should only include public claims

Reply:
i have a test that ensures the d key is not present

Reply:
>    def as_dict(self, is_private=False, **params):

Reply:
>>> set.keys[0].as_dict()
{'e': 'AQAB', 'n': '0bml-h4oJEkmonIBzKZWKoaEt_jn5exY06RwOY-EB6Xp5RPbnQj8AdEW6tl8XBnpdJzhYMb7dnySzRj--jMxG_K8ZhTjLG68og4sm66H08QhWUey1lCN3vTvni8tCZtc7iPKgXJXTzyIkOse-UVkZwhQngPCh7MWjFcG4UfF97APl8XKcjpyshKakfYSpfbKoFqvRbqlJAKCyiwnVf3Ea-RXjh9spLsd77qTNMJEQt14PJxruYXTHPPubKvTJRqwR_ObxYrFxE5h8UZLFk8QYd6k_qKXdV0h2KNuu-PzmyIq7RmQMTr7M4xWexLzrQ7msnsPFJHncXfUD1-jQMtK5Q', 'kty': 'RSA', 'kid': '3eL_LSEgEHCNa45mwU7zZ83SEHvX2MesdKWcP14jQ8s'}

Reply:
perfect, thanks